vm.js

```diff
@@ -1472,34 +1568,59 @@ Object.subclass('Squeak.Object',
         this.hash = hsh;
         this.bits = data;
     },
-    classNameFromImage: function(oopMap) {
-        var name = oopMap[this.bits[Squeak.Class_name]];
-        if (name && name.format >= 8 && name.format < 12) {
-            var bytes = name.decodeBytes(name.bits.length, name.bits, 0, name.format & 3);
-            return Squeak.bytesAsString(bytes);
-        }
-        return "Class";
-    },
-    renameFromImage: function(oopMap, ccArray) {
-        var classObj = this.sqClass < 32 ? oopMap[ccArray[this.sqClass-1]] : oopMap[this.sqClass];
-        var instProto = classObj.instProto || classObj.classInstProto(classObj.classNameFromImage(oopMap));
-        if (!instProto) return this;
-        var renamedObj = new instProto; // Squeak.Object
-        renamedObj.oop = this.oop;
-        renamedObj.sqClass = this.sqClass;
-        renamedObj.format = this.format;
-        renamedObj.hash = this.hash;
-        renamedObj.bits = this.bits;
-        return renamedObj;
-    },
-    installFromImage: function(oopMap, ccArray, floatClass, littleEndian, nativeFloats) {
+    classNameFromImage: function(instanceBits, oopMap) {
+        if((!this.bits || this.bits.length < 6) && (!this.pointers || this.pointers.length < 6) || this.format >= 5)
+            debugger;
+        if(this.bits && this.bits.length > Squeak.Class_name) {
+            var oop = this.bits[Squeak.Class_name];
+            if ((oop & 1) !== 1) {
+                var prop = Object.getOwnPropertyDescriptor(oopMap, oop);
+                if(!prop.get) {
+                    var nameOrCls = prop.value;
+                    if(nameOrCls.format >= 8 && nameOrCls.format < 12) {
+                        if(nameOrCls.bytes)
+                            return Squeak.bytesAsString(nameOrCls.bytes);
+                        var bytes = this.decodeBytes(nameOrCls.bits.length, nameOrCls.bits, 0, nameOrCls.format & 3);
+                        return Squeak.bytesAsString(bytes);
+                    }
+                }
+            }
+        }
+        if(this.pointers && this.pointers > Squeak.Class_name) {
```

Comment:
Missing `.length`. Also, this seems way more complicated than needed - if we have `instanceBits`, why not always use them? In my version, `classNameFromImage` was 5 lines of code. We shouldn't make this more complex than necessary.